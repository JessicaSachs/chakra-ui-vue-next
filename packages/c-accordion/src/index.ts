import { h, defineComponent, PropType } from 'vue'
import { chakra, DOMElements } from '@chakra-ui/vue-system'

const CAccordion = defineComponent({
  props: {
    as: {
      type: [String] as PropType<DOMElements>,
      default: 'div',
    },
  },
  setup(props, { slots, attrs }) {
    return () => h(chakra(props.as), { ...attrs }, slots)
  },
})

export default CAccordion
