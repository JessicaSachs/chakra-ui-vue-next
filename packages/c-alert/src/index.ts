import { h, defineComponent, Component } from 'vue'

const CAlert = defineComponent({
  props: {
    as: {
      type: [String, Object],
      default: 'div',
    },
  },
  render() {
    return h(
      this.as,
      {
        ...this.$attrs,
        role: 'alert',
      },
      this.$slots.default && this.$slots.default()
    )
  },
})

export default CAlert
