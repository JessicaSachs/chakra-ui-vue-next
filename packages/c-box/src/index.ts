import { h, defineComponent, PropType } from 'vue'

const CBox = defineComponent({
  props: {
    as: {
      type: String as PropType<string>,
      default: 'div',
    },
  },
  render() {
    return h(this?.as, { ...this.$props, ...this.$attrs }, this.$slots)
  },
})

export default CBox
