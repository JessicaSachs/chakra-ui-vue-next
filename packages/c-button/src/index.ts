import { h, defineComponent, PropType } from 'vue'

const CButton = defineComponent({
  name: 'CButton',
  props: {
    as: {
      type: String as PropType<string>,
      default: 'button',
    },
  },
  render() {
    return h(
      this?.as,
      { ...this.$props, ...this.$attrs },
      this.$slots.default && this.$slots.default()
    )
  },
})

export default CButton
