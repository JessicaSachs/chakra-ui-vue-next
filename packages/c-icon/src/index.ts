export * from './icon'
export { default as internalIcons } from './icon.internals'
export * from './icon.internals'
