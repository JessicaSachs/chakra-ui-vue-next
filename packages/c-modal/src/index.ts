export { default as CModal } from './c-modal'
