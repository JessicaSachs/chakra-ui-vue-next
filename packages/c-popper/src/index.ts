export { default as usePopper } from './vue-popper'
