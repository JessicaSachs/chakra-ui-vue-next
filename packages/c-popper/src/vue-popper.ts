import { Component, computed, Ref, ref, VNode, watchEffect } from 'vue'
import {
  createPopper,
  VirtualElement,
  Options as PopperOptions,
  Modifier,
  Instance,
} from '@popperjs/core'

export interface Options extends Partial<PopperOptions> {
  enabled: boolean
}

const EMPTY_MODIFIERS: any[] = []

/** Popper state */
type PopperState = {
  styles: {
    [key: string]: Partial<CSSStyleDeclaration>
  }
  attributes: {
    [key: string]: { [key: string]: string | boolean }
  }
}
function usePopper(
  referenceElement: Element | VirtualElement,
  popperElement: HTMLElement,
  /** options should be reactive */
  options: Options
) {
  const optionsWithDefaults = computed(() => ({
    onFirstUpdate: options.onFirstUpdate,
    placement: options.placement || 'bottom',
    strategy: options.strategy || 'absolute',
    modifiers: options.modifiers || EMPTY_MODIFIERS,
  }))

  const styles = ref<PopperState['styles']>({
    popper: {
      position: optionsWithDefaults.value.strategy,
      left: '0',
      top: '0',
    },
  })

  const attributes = ref<PopperState['attributes']>({})

  const updateStateModifier = computed<Modifier<'updateState', any>>(() => ({
    name: 'updateState',
    enabled: true,
    phase: 'write',
    fn: ({ state }) => {
      styles.value = state.styles
      attributes.value = state.attributes
    },
    requires: ['computeStyles'],
  }))

  const popperOptions = computed(() => ({
    ...optionsWithDefaults.value,
    modifiers: [
      ...optionsWithDefaults.value.modifiers,
      updateStateModifier.value,
      { name: 'applyStyles', enabled: true },
    ],
  }))

  // Define popperInstance
  const popperInstance = ref<Instance | null>(null)

  // Set popperOptions to popperInstance
  // when popperOptions change
  watchEffect(
    () => {
      if (popperInstance.value) {
        popperInstance.value.setOptions(popperOptions.value)
      }
    },
    {
      flush: 'sync',
    }
  )

  watchEffect(
    (onInvalidate) => {
      popperInstance.value?.forceUpdate()
      if (!options.enabled || !referenceElement || !popperElement) {
        return
      }

      popperInstance.value = createPopper(
        referenceElement,
        popperElement,
        popperOptions.value
      )

      popperInstance.value?.forceUpdate()

      onInvalidate(() => {
        popperInstance.value?.destroy()
        popperInstance.value = null
      })
    },
    {
      flush: 'sync',
    }
  )

  watchEffect(
    (onInvalidate) => {
      const id = requestAnimationFrame(() => {
        popperInstance.value?.forceUpdate()
      })

      onInvalidate(() => {
        cancelAnimationFrame(id)
      })
    },
    {
      flush: 'sync',
    }
  )

  return {
    ...popperInstance.value,
    state: popperInstance.value?.state || null,
    styles,
    attributes,
    update: popperInstance.value?.update || null,
    forceUpdate: popperInstance.value?.forceUpdate || null,
  }
}

export default usePopper
