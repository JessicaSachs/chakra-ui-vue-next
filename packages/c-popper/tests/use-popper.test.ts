import { render, waitMs } from '@chakra-ui/vue-test-utils/src'
import { usePopper } from '../src'
import { useToggle } from '@vueuse/core'
import { CButton } from '@chakra-ui/c-button'
import { onMounted } from 'vue'

const renderComponent = (props?: any) => {
  const base = {
    components: {
      CButton,
    },
    template: `
      <div>
        <c-button :ref="reference" @click="toggleIsOpen">Testing</c-button>
        <div v-if="isOpen" :ref="popper" style="padding: 20px; background: red">
          <div
            data-popper-arrow=""
            style="--popper-arrow-size: 10px; background: yellow"
          />
          Popper
        </div>
      </div>
    `,
    ...props,
  }
  return render(base)
}

it('should render properly', async () => {
  let _referenceEl: any
  let _popperEl: any
  const { asFragment } = renderComponent({
    setup() {
      const [isOpen, toggleIsOpen] = useToggle(true)

      const { reference, referenceEl, popperEl, popper } = usePopper({
        gutter: 16,
        placement: 'right-end',
      })

      onMounted(() => {
        _referenceEl = referenceEl.value
        _popperEl = popperEl.value
      })

      return {
        isOpen,
        toggleIsOpen,
        reference,
        popper,
      }
    },
  })

  await waitMs(300)
  expect(asFragment()).toMatchSnapshot()
  expect(_referenceEl instanceof Element).toBe(true)
  expect(_popperEl instanceof Element).toBe(true)
})
