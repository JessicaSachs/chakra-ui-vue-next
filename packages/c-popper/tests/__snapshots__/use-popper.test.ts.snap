// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<DocumentFragment>
  <div>
    <button
      class="chakra-button css-myeq6i"
    >
      <!---->
      <!---->
      Testing
      <!---->
    </button>
    <div
      data-popper-placement="right-end"
      style="padding: 20px; background: red; --popper-transform-origin: left bottom; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(16px, 0px);"
    >
      <div
        data-popper-arrow=""
        style="--popper-arrow-size: 10px; background: yellow; z-index: -1; --popper-arrow-size-half: calc(var(--popper-arrow-size, 8px) / 2); --popper-arrow-offset: calc(var(--popper-arrow-size-half) * -1); position: absolute; top: 0px; transform: translate(0px, 8px);"
      />
       Popper 
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render properly and bind refs 1`] = `
<DocumentFragment>
  <div>
    <button
      class="chakra-button css-myeq6i"
    >
      <!---->
      <!---->
      Testing
      <!---->
    </button>
    <div
      data-popper-placement="right-end"
      style="padding: 20px; background: red; --popper-transform-origin: left bottom; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(16px, 0px);"
    >
      <div
        data-popper-arrow=""
        style="--popper-arrow-size: 10px; background: yellow; z-index: -1; --popper-arrow-size-half: calc(var(--popper-arrow-size, 8px) / 2); --popper-arrow-offset: calc(var(--popper-arrow-size-half) * -1); position: absolute; top: 0px; transform: translate(0px, 8px);"
      />
       Popper 
    </div>
  </div>
</DocumentFragment>
`;
